### ISSUE TITLE
 Long Station Patch

### ISSUE THREAD

**Issue by ERRORTEK23** [OpenRCT2/OpenLoco#126]
Another patch, just wrong, can not use the aircraft.

https://www.tt-forums.net/viewtopic.php?f=38&t=50627

**Comment by duncanspumpkin** [comment]
There is absolutely no information in this issue. Are you saying that you want this in openloco? Are you saying that it doesn't work in openloco?

**Comment by AaronVanGeffen** [comment]
I'm guessing it's about the limits. According to the thread, after applying the patch:
> you are now able to build big stations with 243 tiles instead of the regular 80 tiles.

**Comment by ERRORTEK23** [comment]
It would be great if it was in openloco. Because there are only 80 tiles in the original game. And this patch does not work under openloco. I thought it would help with programming.

**Comment by jgottula** [comment]
I can potentially implement this basically right now, in a similar manner to what I did in #25. (If that's desired, and not objectionable.)

I'll see if I can get a solid grip on how many patch locations there are, and what they are, and where they are.

**Comment by AaronVanGeffen** [comment]
If it's not too many patch locations, I'd rather the functions themselves were implemented rather than patched around…

**Comment by jgottula** [comment]
Actually, and this is a bit weird, the README file for that patch seems to suggest that title.dat needs to be overwritten with a zero-byte file, for some reason. With this corresponding warning:

> In the main menu an error message will show up that the background demo cannot be loaded. This is normal and not to handle yet.

And it also says it breaks savegame compatibility. Apparently in such a manner that you can't load *old* savegames in the patched version of the game, and not the other way around? Did I read this correctly?

> Old savegames are not yet compatible, a savegame converter is under development

----

I'll maybe take a look and see if this is all due to them doing the patch in some sort of boneheaded way, when there may be a better way to go about it that doesn't involve such large limitations.

Also note that if we implement this down the road after things have been decompiled more completely, then the title situation would almost certainly not be a problem, and the savegame compatibility thing may or may not be an actual problem depending on the details there.

**Comment by jgottula** [comment]
> If it's not too many patch locations, I'd rather the functions themselves were implemented rather than patched around…

Of course. I'll see about tracking down the patch offsets and determine if they're clustered in easy-to-implement functions.

**Comment by jgottula** [comment]
Looks like 270 individual patch locations.

- 186 cases of doubling an immediate value from `0x000003D2` to `0x000007A4`
- 10 cases of halving an immediate value from `0x00000400` to `0x00000200`
- 1 case of changing an immediate value from `0x0100` to `0x0500`
- 1 case of changing an immediate value from `0x000000f0` to `0x000002d9`
- 8 cases of changing a comparison with `0x0050` into a `call` + `nop`s
  - (incidentally, it seems that the `word` at `station+0x1ce` holds the number of tiles that the station's catchment area covers)
- 1 case of tacking on a new function at the end of `.text` (which I presume is the call target of the 8 cases mentioned above), which consists of `cmp word ptr [ebx+0x5e70aa],0xf3; ret`
- 62 cases of adding `sizeof(station)` to an immediate address reference (i.e. making the instruction refer to `&stations[0]` rather than `&stations[1]` as the base of the array)
- 1 case of changing an immediate offset from `0x000003ba` to `0x000008cb` (probably a combination of bumping up the station array index and changing the field being modified by the instruction)

So... yeah there's *a lot of stuff* in there. Plenty more than my window/viewport patch had. And I have no idea if what they're doing is sane at all. The apparent station array index manipulation in particular doesn't make any sense to me and seems like a really bad idea. But again, I have no idea what they're accomplishing by doing that.

> I could potentially implement this basically right now

^ I hereby rescind this statement that I made earlier, as I said it before I realized that this involved 270 independent patches, many of which seem to be doing kinda sketchy/non-straightforward things. :stuck_out_tongue:

**Comment by marijnvdwerf** [comment]
These are the functions it appears to affect:

```
sub_4267BE
sub_426B29
sub_426D52
sub_426E26
sub_426F0B
sub_427122
sub_427214
sub_4273DF
sub_42773C
sub_427FC9
sub_428379
sub_4285BA
sub_429209
sub_42A036
sub_42F2BF
sub_42F2FE
sub_42F489
sub_42F550
sub_4308D4
do_game_command
sub_455943
sub_4574E8
sub_46C0AE
sub_46CD31
sub_47036E
sub_470B76
sub_475FBC
sub_4ACEE7
sub_47DFD0
sub_482662
sub_48667A
sub_487784
sub_4883D4
sub_48B1D8
sub_48B1FA
towns_update_daily_48B244
paint_stations
sub_48BB20
sub_48C402
sub_48C708
sub_48D2AC
sub_48DDC3
sub_48DDE4
sub_48DE97
sub_48E32C
ui::station::event_1_prepare_draw
ui::station::event_1_text_input
ui::station::event_2_prepare_draw
ui::station::event_2_draw
ui::station::event_2_draw_scroll
ui::station::event_2_get_scroll_size
ui::station::event_3_prepare_draw
ui::station::event_3_draw_scroll
ui::station::event_3_get_scroll_size
sub_48EFBC
sub_48F11B
window_station_create
sub_48F321
sub_48F43A
sub_48F482
sub_48F529
sub_48F716
sub_48F7D1
sub_48F850
sub_48F8A0
sub_48F988
sub_48FEF4
sub_4900B8
sub_490372
sub_490756
sub_49088B
sub_4910AB
sub_4910E8
sub_49111A
sub_4911FD
sub_491247
sub_491281
sub_4912BB
ui::station_list::event_5_draw_scroll
sub_492640
sub_49271A
update_station_cargo
sub_4929DB
sub_492A98
sub_492C41
sub_492F3D
sub_493559
sub_4937FA
sub_493AA7
sub_494570
sub_494805
sub_49711F
sub_49B4E0
ui::construction::event_2_draw
ui::construction::event_2_10
sub_4A734F
vehicle_4A8B81
sub_4AC3D3
ui::vehicle::event_3_draw_scroll
sub_4B49F8
ui::vehicle::event_5_scroll_mouse_down
sub_4B5A1A
sub_4B677B
sub_4B67A9
sub_4B67D7
sub_4B6805
sub_4B682F
sub_4B6859
sub_4B6885
sub_4B996F
sub_4B9A2A
sub_4BA142
ui::vehicle_list::event_6_scroll_mouse_over
process_mouse_over
viewport_interaction_get_item_left
viewport_interaction_right_over
```
